src/afrolabs/components/kafka.clj

```diff
@@ -1750,76 +1892,206 @@
-                            ;; the encountered offset with. Choose `-1` because it is less than
-                            ;; any offset we will encounter (ie invalid offset value) but still
-                            ;; an int that will not throw inside max.
-                            (and o p)
-                            (update-in [:ktable/topic-partition-offsets t p]
-                                       (fnil #(max % o)
-                                             -1))
-
-                            hdrs
-                            (assoc-in [:ktable/record-headers t k]
-                                      hdrs)
-
-                            (not hdrs)
-                            (update-in [:ktable/record-headers t]
-                                       #(dissoc % k)))))
-             rest-msgs))))
+  "Acts like a reduce function.
+  Accepts a starting-state (`old-ktable`) and a collection of kafka-records (`new-msgs`),
+  Merges the `new-messages` into a map, which is returned.
+  The state has the shape {topic-name -> {record-key latest-record-value}}.
+  Each `latest-record-value` has meta-data including `:headers` & `:timestamp`.
+  The result value (state) has meta-data
+  - `:ktable/topic-partition-offsets`
+  - `:ktable/record-headers`
+  that records the max offset, per [topic partition] encountered while building the ktable value.
+  Accepts an optional `opts` parameter:
+  - :retention-ms -- If a `latest-record-value` was created from a kafka-record wit a timestamp
+  thas is older than latest-received-kafka-record-timestamp, then this record will be removed from the ktable."
+  ([old-ktable new-msgs]
+   (merge-updates-with-ktable old-ktable new-msgs {}))
+  ([old-ktable new-msgs {:as   _opts
+                         :keys [retention-ms]}]
+   (let [[result max-timestamp]
+         (loop [old                           old-ktable
+                [{:as  head
+                  k    :key
+                  v    :value
+                  t    :topic
+                  p    :partition
+                  o    :offset
+                  ts   :timestamp
+                  hdrs :headers} & rest-msgs] new-msgs
+                max-timestamp                 ts]
+           (let [new-max-timestamp (max-instant max-timestamp ts)]
+             (if-not head
+               [old max-timestamp]
+               (recur (vary-meta (let [v-meta-data (cond-> {}
+                                                     hdrs (assoc :headers hdrs)
+                                                     ts   (assoc :timestamp ts)
+                                                     p    (assoc :partition p)
+                                                     o    (assoc :offset o))
+                                       v (if (and v (instance? clojure.lang.IMeta v))
+                                           (with-meta v v-meta-data)
+                                           v)]
```

Comment:
keep all of the record meta-data (and not only the `:timestamp`). Also, check that the record-value can actually carry meta-data before trying to assign it.